include/clink/jna/jna_context.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2021 The Clink Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CLINK_JNA_JNA_CONTEXT
+#define CLINK_JNA_JNA_CONTEXT
+
+#include "clink/utils/clink_utils.h"
+#include "tfrt/bef_executor_driver/bef_executor_driver.h"
+#include "tfrt/data/dataset.h"
+#include "tfrt/host_context/execution_context.h"
+#include "tfrt/support/ref_count.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+
+namespace clink {
+
+// JnaContext holds the context information for kernel and op execution, which
+// currently includes the host context.
+//
+// JnaContext is only supposed to be used in functions for JNA cases. It is
+// recommended to create a global static instance of JnaContext and to reuse the
+// instance across all JNA functions in the same memory space.
+class JnaContext {
+public:
+  explicit JnaContext()
```

Comment:
Would it be simpler to use `tfrt::HostContext` directly instead of creating the `JnaContext` class?

Reply:
`JnaContext` was supposed to include all objects that might be needed by JNA functions. Like `RequestContext` has `HostContext`, `ResourceContext` and `ContextData`, I wondered whether JNA also needs objects other than a single `HostContext`. If we need to add such classes in future, we can just add the variable and corresponding methods to `JnaContext`, and all JNA functions that has been using `JnaContext` can remain untouched. Maybe I'm just a little bit overly worried and over-wrapped the classes.

Reply:
In general every indirection/abstraction adds development complexity as well as runtime overhead. So we need to have a good reason for why we need this. This reason could be based on a plan that we will implement in the future.

If we don't have a clear idea of why we need something, it seems simper to start simple and avoid over-enginneering.

Reply:
OK. I'll remove `JnaContext`. Besides, below is another problem I encountered together with `JnaContext`. I have found my own solution, but it might still worth a look.

You may have noticed it, that I created the `HostContext` with something like this:
```c++
host_ = CreateHostContext("mstd", tfrt::HostAllocatorType::kMalloc).release();
```
This is because smart pointers like `unique_ptr` malfunctions when it comes to JNA, when the pointers seem unable to compute the number of reference to this pointer correctly and would accidentally free the object. Thus I called `release()` to make it a normal C pointer. Besides, as we are about to remove `JnaContext` and there would only be a static `HostContext` for all JNA functions, I think we don't need to worry about memory leak due to this pointer.